app/services/notifications/mentioned_in_post_service.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Notifications
+  class MentionedInPostService
+    def self.model
+      Notifications::MentionedInPost
+    end
+
+    def self.notify(mentionable, recipient_user_ids)
+      actor = mentionable.author
+      relevant_mentions = filter_mentions(
+        mentionable.mentions.local.where.not(person: actor),
+        mentionable,
+        recipient_user_ids
+      )
+
+      relevant_mentions.each do |mention|
+        recipient = mention.person.owner
+        next if model.exists?(recipient: recipient, target: mention)
```

Comment:
This relies on the notification entity existing, which isn't the case if in-app notifications are disabled. This was a workaround for when posts were received twice, no duplicate notifications were created (https://github.com/diaspora/diaspora/pull/7721). I don't remember all the details and reasons, but `Diaspora::Federation::Receive.status_message` still processes a received post even, if it already exists, and because of that it also calls `ReceiveLocal` again.

I don't know/remember why a post should be received again if it already exists, maybe that was a workaround for when a post was created incomplete because of the scenario described [here](https://discourse.diasporafoundation.org/t/federation-issue-trigger-again/4648/4?u=supertux88) where some data for posts was missing, and receiving them again properly through the federation, would add this missing data. But if that gets fixed so it fetches the posts using federation code, it doesn't need to be received a second time, as it was already complete the first time. That would then also solve the problem of duplicate notifications, without needing the check for them here (as I don't know how otherwise this should be solved for not sending duplicate mails).

Reply:
@SuperTux88 is it enough to do the proper fetching using federation code? It seems quite easy.

https://github.com/diaspora/diaspora/commit/880b2585ffbb231951eacec73df5a460db959517

Didn't look at the tests yet.

Reply:
(Probably worth adding exception handling too)

Reply:
> is it enough to do the proper fetching using federation code? It seems quite easy.

Yes, that's theoretically all that's needed, but the `type` should be the lowercase, snake_case version of the entity name ([doc](https://github.com/diaspora/diaspora_federation/blob/develop/lib/diaspora_federation/federation/fetcher.rb#L9)).

But you can even improve that a little bit more by just using `:post` as type (as it could also be a reshare) and you can remove the `check_type` in `validate`, so that way we can even fetch reshares (you could also use the type from the json, but then you would need to convert that to the snake_case version, so just always use `:post` works and is easier) :)

But yes, basically that, it's fairly easy :+1:

Reply:
And the exception should be `DiasporaFederation::Federation::Fetcher::NotFetchable`.